### ISSUE TITLE
segfault in pmix_ptl_base_lost_connection()

### ISSUE THREAD

**Issue by krempel-pt** [pmix/pmix#1122]
### What version of the PMIx Reference Library are you using?
v3.1.2

### Describe how PMIx was installed
Compiled from the release tarball using 
configure --disable-static --enable-shared --enable-pmi-backward-compatibility --with-munge

### Please describe the system on which you are running
* Operating system/version: CentOS Linux release 7.5.1804 (Core)
* Computer hardware: Intel Xeon E5-2650
* Network type: InfiniBand

-----------------------------

## Details of the problem
I'm implementing PMIx support into a resource manager using the server library provided by the pmix reference implementation. In the case of some processes fail for some reason, I occasionally see my pmix server process segfault in the pmix library:
```gdb
Program terminated with signal 11, Segmentation fault.
#0  0x00007fa0137d33b2 in pmix_ptl_base_lost_connection (peer=0x7fa008030e90, err=-25) at base/ptl_base_sendrecv.c:109                                                    
109                     if (!PMIX_CHECK_PROCID(&rinfo->peer->info->pname, &peer->info->pname)) {                                                                          
(gdb) bt
#0  0x00007fa0137d33b2 in pmix_ptl_base_lost_connection (peer=0x7fa008030e90, err=-25) at base/ptl_base_sendrecv.c:109                                                    
#1  0x00007fa0137d6a89 in pmix_ptl_base_recv_handler (sd=<optimized out>, flags=<optimized out>, cbdata=0x7fa008030e90) at base/ptl_base_sendrecv.c:628                   
#2  0x00007fa0132bba14 in event_base_loop () from /usr/lib64/libevent-2.0.so.5
#3  0x00007fa01378257e in progress_engine (obj=<optimized out>) at runtime/pmix_progress_threads.c:109                                                                    
#4  0x00007fa01a490e25 in start_thread () from /usr/lib64/libpthread.so.0
#5  0x00007fa01b514bad in clone () from /usr/lib64/libc.so.6
(gdb) print *rinfo
$1 = {super = {super = {obj_class = 0x3348640824548b48, obj_reference_count = 2630932}, pmix_list_next = 0xcbe890c318c48348, pmix_list_prev = 0x1f0f2e6666fffe6c,         
    item_free = 132}, ev = {ev_active_next = {tqe_next = 0x2825048b486418ec, tqe_prev = 0x824448948000000}, ev_next = {tqe_next = 0x24348904c1f6c031,                     
      tqe_prev = 0x2444c60424548966}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x7400062444c60007, tqe_prev = 0xc88320e083c8890c},                     
      min_heap_idx = 1153826823}, ev_fd = 608471041, ev_base = 0x1c98320e1830a74, _ev = {ev_io = {ev_io_next = {tqe_next = 0x87f8b4806244c88,                             
          tqe_prev = 0xe80cc78348e68948}, ev_timeout = {tv_sec = 2617870429057972730, tv_usec = 11299767758382088}}, ev_signal = {ev_signal_next = {                      
          tqe_next = 0x87f8b4806244c88, tqe_prev = 0xe80cc78348e68948}, ev_ncalls = -1542, ev_pncalls = 0x28251433486408}}, ev_events = 0, ev_res = 1397,                 
    ev_flags = -31928, ev_pri = 196 '\304', ev_closure = 24 '\030', ev_timeout = {tv_sec = 5469621748584353731, tv_usec = -8926399244454068628},                          
    ev_callback = 0x5441900000000000, ev_arg = 0x10ec834853fd8955}, event_active = 100, trk = 0xc031082444894800, hdr = {pindex = -26693144, tag = 1032014079,            
    nbytes = 14089872074044746065}, peer = 0xe33d8b4174208b44}
(gdb) print rinfo->peer
$2 = (pmix_peer_t *) 0xe33d8b4174208b44
(gdb) print *rinfo->peer
Cannot access memory at address 0xe33d8b4174208b44
```

**Comment by rhc54** [comment]
We have been working on resolving race conditions in that area of the code, but I confess this is a new one to me. Can you provide any details on what the client processes were doing at the time one or more failed? Were they engaged in some PMIx collective op like "fence"? Did multiple procs fail or just one? Etc. Any further info would be helpful.

**Comment by krempel-pt** [comment]
Unfortunately I cannot give you much more information, since reproducing of the problem is hard since it occurred in situations that I tried to avoid.
As far as I can say, the clients were stuck for some reasons in a fence operation and the segfault happened in the server during or after terminating the clients by sending them a SIGTERM.

**Comment by rhc54** [comment]
No need - you told me exactly what I needed to know. I was chasing down the same issue this weekend and have a fix that I will backport to the release branches later this week.

**Comment by wlepera** [comment]
I'm seeing a similar issue on ppc64le, using 3.1.2 PMIx, where an application terminates with error and orted segfaults with a similar (but not identical) backtrace:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x20000146f1b0 (LWP 52737)]
0x0000200000c6f2b8 in pmix_ptl_base_lost_connection () from /smpi_dev/lepera/urial/ibm-smpi/ompibase/exports/debug/lib/libpmix.so.2
(gdb) where
#0  0x0000200000c6f2b8 in pmix_ptl_base_lost_connection () from /smpi_dev/lepera/urial/ibm-smpi/ompibase/exports/debug/lib/libpmix.so.2
#1  0x0000200000c72608 in pmix_ptl_base_recv_handler () from /smpi_dev/lepera/urial/ibm-smpi/ompibase/exports/debug/lib/libpmix.so.2
#2  0x00002000005ab854 in event_persist_closure (ev=<optimized out>, base=<optimized out>) at event.c:1580
#3  event_process_active_single_queue (base=0x4b8330e0, activeq=0x4b8335b0, max_to_process=2147483647, endtime=0x0) at event.c:1639
#4  0x00002000005ac404 in event_process_active (base=0x4b8330e0) at event.c:1738
#5  event_base_loop (base=0x4b8330e0, flags=<optimized out>) at event.c:1961
#6  0x0000200000c0aeec in progress_engine () from /smpi_dev/lepera/urial/ibm-smpi/ompibase/exports/debug/lib/libpmix.so.2
#7  0x0000200000638b94 in start_thread () from /lib64/libpthread.so.0
#8  0x00002000007985f4 in clone () from /lib64/libc.so.6
(gdb) 
```
Is this fix available in v3.1 branch?

**Comment by jjhursey** [comment]
@rhc54 Did https://github.com/pmix/pmix/pull/1140 target this defect or was there something more?

**Comment by rhc54** [comment]
It should fix this, but I'd like to see it verified if possible

**Comment by rhc54** [comment]
@krempel-pt @wlepera Would it be possible for you to test PR https://github.com/pmix/pmix/pull/1162 ? I believe it should fix the problems you observed.

**Comment by wlepera** [comment]
I applied #1140 first and still saw the problem, but adding #1162 seems to resolve my issue, thank you.  Just to note for the future, the original symptom reported to us was when an application running across two compute nodes terminated with error, the user observed an infinite loop of "mpirun: Forwarding signal 11 to job" messages streamed to the console.  After we killed the run with ctrl-c we got the core file with the stack trace referenced above.

**Comment by rhc54** [comment]
Thanks for the details - I'll pass that along to the OMPI folks. Actually, #1162 contains #1140 plus some things that came from other PRs.

**Comment by akesandgren** [comment]
Is there a similar fix for the 2.2 series? We got hit by the same problem (probably) on version 2.2.1

**Comment by rhc54** [comment]
I am working on a new 2.2 release now - it will include these fixes.
